bin/oneoff/data_fix/fix_invalid_users

```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env ruby
+
+require_relative '../../../dashboard/config/environment'
+
+# This script fixes invalid rows in the user table that were added to the database
+# as a result of a bug in https://github.com/code-dot-org/code-dot-org/pull/38026.
+# Briefly, there was a bug in that PR that caused user accounts created using the
+# join section endpoint (FollowersController#student_register) to be incomplete
+# or invalid (i.e. does not pass rails User model validation). The bug was present
+# only in v2020-12-02.0 which was in production from about 12/2/20 2:00pm PST to
+# 12/3/20 noon PST.
+#
+# The rows added to the user table were invalid in one of two ways:
+# 1. If the user passed valid data to the join section endpoint, the inserted row
+#    was valid except that the username field was missing. We fix these by
+#    calling generate_username to generate the username.
+# 2. If the user passed invalid data to the join section endpoint, the inserted row
+#    would contain the invalid data as well as have a missing provider field. We fix
+#    these row by (soft) deleting them.  In many cases, students were not able to
+#    access these accounts. For a small number of cases (n=25), students have
+#    accessed these accounts. For the students who successfully joined a section
+#    (n=9), we will reach out to the teacher to explain the situation.
+
+CDO.log = Logger.new(STDOUT)
+ActiveRecord::Base.record_timestamps = false
+
+options = {actually_update: false}
+OptionParser.new do |opts|
+  opts.banner = "Usage: #{File.basename(__FILE__)} [options]"
+  opts.on('-u', '--actually-update', 'Actually perform the update.') do
+    options[:actually_update] = true
+  end
+  opts.on('-h', '--help', 'Add -u to perform the update.') do
+    puts opts
+    exit
+  end
+end.parse!
+CDO.log.info "Called with options: #{options}"
+
+# To speed up the queries, we limit the query to an id range that we know covers
+# all possible affected users
+START_USER_ID = 71_826_397
+END_USER_ID = 71_929_059
+BATCH_SIZE = 200
+
+changes_committed = 0
+changes_rolled_back = 0
+User.after_commit {changes_committed += 1}
+User.after_rollback {changes_rolled_back += 1}
+
+# Address case 1 - generate username
+usernames_generated = 0
+generate_username_errors = 0
+User.where(username: nil).
+  where.not(provider: nil).
```

Comment:
Do we need to include deleted users here? what will happen if a deleted account in this range is left in a bad state?

Reply:
(because of the paranoia gem, deleted users are only included the query when User.with_deleted is used)

Reply:
It's a good question.  I checked and there are only two users in this group that are deleted and they appear to be Elijah's test accounts.  Also, note that we are not guaranteeing that rows representing deleted users are valid.  All of the Group 2 rows that we are marking as deleted may still be invalid.